File: 1300-critical-connections-in-a-network/critical-connections-in-a-network.js
/**
 * @param {number} n
 * @param {number[][]} connections
 * @return {number[][]}
 */
var criticalConnections = function(n, connections) {
    const disc = new Array(n);
    const low = new Array(n);
    const ans = [];
    const graph = new Array(n);
    let time = 0;
    
    for(let i=0; i<n; i++) {
        graph[i] = [];
    }

    for(con of connections) {
        graph[con[0]].push(con[1]);
        graph[con[1]].push(con[0]);
    }

    function dfs(u, p) {
        low[u] = disc[u] = ++time;

        for(let v of graph[u]) {
            if(v === p) {
                continue;
            }
            if(!disc[v]) {
                dfs(v, u);
                if(disc[u] < low[v]) {
                    ans.push([u, v]);
                }
                low[u] = Math.min(low[u], low[v]);
            }
            else {
                low[u] = Math.min(low[u], disc[v]);
            }
        }
    }

    for(let i=0; i<n; i++) {
        if(!disc[i]) {
            dfs(i, i);
        }
    }

    return ans;
};

File: 1300-critical-connections-in-a-network/critical-connections-in-a-network.java
class Solution {
    int time;
    int[] low;
    int[] disc;
    List<List<Integer>> bridges;
    List<List<Integer>> graph;
    void dfs(int u, int p) {
        low[u] = disc[u] = ++time;
        for(int v : graph.get(u)) {
            if(v == p) continue;
            if(disc[v] == 0) {
                dfs(v, u);
                if(disc[u] < low[v]) {
                    bridges.add(List.of(u, v));
                }
                low[u] = Math.min(low[u], low[v]);
            }
            else {
                low[u] = Math.min(low[u], disc[v]);
            }
        }
    }
    public List<List<Integer>> criticalConnections(int n, List<List<Integer>> connections) {
        low = new int[n];
        disc = new int[n];
        bridges = new ArrayList<>();
        graph = new ArrayList<List<Integer>>(n);
        time = 0;
        for(int i=0; i<n; i++) {
            graph.add(new ArrayList<>());
        }
        for(var con : connections) {
            graph.get(con.get(0)).add(con.get(1));
            graph.get(con.get(1)).add(con.get(0));
        }
        for(int i=0; i<n; i++) {
            if(disc[i] == 0) {
                dfs(i, i);
            }
        }
        return bridges;
    }
}